### PR TITLE
Allow for supplying Filters to Logging Appenders

### PR DESCRIPTION
**Pull request by ghenkes**
This change adds a seam to allow consumers to supply Logback Filters to apply to logging Appenders. These filters will be applied, in order, after the threshold filter. This allows for further refinement of what will be logged. For example, I have an endpoint that is called on a schedule to check availability that I would like to exclude from the access logs by supplying a filter for this URI.

You could also add appenders with the following:
* A filter that will only log failures based on status code.
* A filter that will log requests with slow responses.
* A filter that will log large requests/responses.

### CODE REVIEW

**Comment by arteam** [comment]
This seems to me like a reasonable improvement. It allows to extend logging behavior without changing the default logging configuration. It fits nicely with Dropwizard's philosophy of keeping things simple for the average user, but at the same time providing a means of control for those, who want to customize the framework.

File: dropwizard-logging/src/test/java/io/dropwizard/logging/DefaultLoggingFactoryTest.java [hunk 1]
```diff
@@ -71,6 +71,7 @@ public void canParseNewLoggerFormat() throws Exception {
         assertThat(fileAppenderFactory.getCurrentLogFilename()).isEqualTo("${new_app}.log");
         assertThat(fileAppenderFactory.getArchivedLogFilenamePattern()).isEqualTo("${new_app}-%d.log.gz");
         assertThat(fileAppenderFactory.getArchivedFileCount()).isEqualTo(5);
+        assertThat(fileAppenderFactory.getFilterFactories()).containsExactly(new TestFilterFactory(), new SecondTestFilterFactory());
```
Comment (arteam) [inline dropwizard-logging/src/test/java/io/dropwizard/logging/DefaultLoggingFactoryTest.java]: This currently produces a compiler warning about an unchecked call. What do you think about replacing this
check by a serie of checks?
```java
final ImmutableList<?> filterFactories = fileAppenderFactory.getFilterFactories();
assertThat(filterFactories).hasSize(2);
assertThat(filterFactories.get(0)).isExactlyInstanceOf(TestFilterFactory.class);
assertThat(filterFactories.get(1)).isExactlyInstanceOf(SecondTestFilterFactory.class);
```

As a benefit, we could get rid of `hashCode` and `equals` implementations for `TestFilterFactory` and `SecondTestFilterFactory`.

Comment (arteam) [inline dropwizard-logging/src/test/java/io/dropwizard/logging/DefaultLoggingFactoryTest.java]: After another look, I beleive we could even use `ImmutableList<FilterFactory<ILoggingEvent>>` if we generify `appenderFactory` and `fileAppenderFactory`.

**Comment by arteam** [comment]
@ghenkes, I reviewed the implementation, looks great. One note in a unit test, but generally this looks ready to merge!

**Comment by ghenkes** [comment]
@arteam I've applied your suggestions in the unit test. Thanks!

**Comment by arteam** [comment]
Thanks for your contribution!

**Pull request merged**